Add SOCKS5 proxy support for kubectl exec

TGPSKI opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

The PR permits using kubectl commands like `kubectl exec` via a SOCKS5 proxy. The PR adds support for SOCKS5 proxies, but does not change HTTP or HTTPS behavior.

Test coverage provided for socks5 with [armon/go-socks5](https://github.com/armon/go-socks5) server.

**Does this PR introduce a user-facing change?**:

```release-note
Adds support for kubectl commands (`kubectl exec`) via a SOCKS5 proxy. 
```

TGPSKI commented:
I signed it

TGPSKI commented:
/check-cla

TGPSKI commented:
/check-cla

TGPSKI commented:
/check-cla

TGPSKI commented:
I signed it

TGPSKI commented:
@cheftako @mikedanese I've resurrected the [SOCKS5 Proxy PR](https://github.com/kubernetes/kubernetes/pull/72943), updated and fixed the code, and have added test coverage. 

I'm wondering if you could take a look & review?

TGPSKI commented:
/assign @lavalamp

fedebongio commented:
/assign @mikedanese @caesarxuchao

TGPSKI commented:
/assign @smarterclayton @liggitt @deads2k

lavalamp commented:
/ok-to-test

The code seems pretty straightforward; @mikedanese and/or @liggitt are there any security implications of accepting a new proxy mechanism?

How many people are in the position where they need to use this? We need to evaluate potential usage vs. ongoing maintenance cost.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -144,7 +157,7 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
 	}
 
-	proxyDialConn, err := s.dialWithoutProxy(req.Context(), proxyURL)
+	proxyDialConn, err := s.dialWithoutProxy(proxyReq.Context(), proxyURL)
```
`proxyReq.Context()` is detached from the original req.Context(), which means we lose timeout/cancellation threading

TGPSKI replied:
Passed req into dialWithHttpProxy and used req.WithContext to assume the original request context.

TGPSKI commented:
Does anyone have comments on fixing a `Forbidden - CSRF token invalid` error during the test process? I can't find information about it with standard searching.

TGPSKI commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -22,13 +22,17 @@ import (
 	"testing"
+	"time"
 
+	"github.com/armon/go-socks5"
```
I thought this might be a concern. It's just for test coverage, but it's not ideal. Other packages providing a socks5 server don't look great either. If there's not a suitable test server, I can't include test coverage. What is an acceptable course of action here?

https://github.com/dddpaul/go-socks5-server - 13 commits
https://github.com/ziozzang/socks5-proxy - 9 commits
https://github.com/ring04h/s5.go - 5 commits

liggitt replied:
The lack of activity is not ideal, but there are no non-stdlib transitive dependencies, the repo is small, has an acceptable license, and no churn. I don't see a better alternative, and we wouldn't want to add untested function.

TGPSKI commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

TGPSKI commented:
@lavalamp @liggitt Happy new year!

I fixed the request context issue raised, and also included the state of SOCKS5 test proxy servers (only three, not well maintained repos) https://github.com/kubernetes/kubernetes/pull/84205#discussion_r342363339.

I need advice on where to go next:

1. Do we include the SOCKS5 proxy test server, even though the packages supporting it are not maintained?
2. If we don't include a proxy test server, are we OK with running the SOCKS5 proxy without tests?
3. Where can I find a resource on adding a module to the k8s repo? I have spent a lot of time and have failed to correctly add the SOCKS5 proxy server module for testing. Is there a clear doc showing a walkthrough that I am missing?

kfox1111 commented:
Do we need to test with mock or would an integration test be enough? ssh has built in socks5 support so maybe that could be used for an integration test?

This feature would be very useful.

TGPSKI commented:
@kfox111 its not just a mock, its actually an integration test with a go-socks5 server.

https://github.com/kubernetes/kubernetes/pull/84205/files#diff-1a90f55d4557529033c309d1103a40fcR76

kfox1111 commented:
Ah, yeah. That does a bit more testing then what ssh could do. but maybe doing it with ssh without the proxyauth tests is better then no tests at all if the go socks5 lib is unacceptable?

I think all the e2e environments have access to ssh presetup? Can the node ssh to itself? if so, you could 'ssh -D 10001 localhost sleep 10000 & and then try the test with the proxy set to localhost:10001.

sftim reviewed (changes requested):
The suggested release note mentions a pull request (PR). I recommend rewriting it to look like a fragment of the notes for a Kubernetes release.

TGPSKI commented:
> The suggested release note mentions a pull request (PR). I recommend rewriting it to look like a fragment of the notes for a Kubernetes release.

Thanks for the feedback.

Do you have any thoughts on this comment? https://github.com/kubernetes/kubernetes/pull/84205#discussion_r342363339

I have been stalled on this sticking point since I submitted the PR.

RA489 commented:
/check-cla

TGPSKI commented:
I haven't gotten any updates on the three open questions I posed a while ago. I have 3 open questions on this PR & need some direction to move forward.

1. Do we include the SOCKS5 proxy test server, even though the packages supporting it are not maintained?
2. If we don't include a proxy test server, are we OK with running the SOCKS5 proxy without tests?
3. Where can I find a resource on adding a module to the k8s repo? I have spent a lot of time and have failed to correctly add the SOCKS5 proxy server module for testing. Is there a clear doc showing a walkthrough that I am missing?

kfox1111 commented:
+1. How do we make progress on this? I have some users that would greatly benefit from this.

lavalamp commented:
At this point, I think you should try bugging @mikedanese and/or @liggitt on slack?

zioalex commented:
Hi guys, can we push this one? 
It would open the great possibility to interact with Kubernetes behind a proxy...That in Corporate environments nowadays can be very useful.

TGPSKI commented:
> Hi guys, can we push this one?
> It would open the great possibility to interact with Kubernetes behind a proxy...That in Corporate environments nowadays can be very useful.

I'm totally with you! Still need answers to the open questions from the maintainers. They have been AFK on this one for months.

erikbryant commented:
> At this point, I think you should try bugging @mikedanese and/or @liggitt on slack?

We would really like to get this merged. What can we do to get it finalized?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -157,7 +172,46 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+			Password: pass,
+		}
+	}
+	proxyDialer, err := proxy.SOCKS5("tcp", proxyDialAddr, auth, proxy.Direct)
```
should this use `s.Dialer` if it is non-nil, rather than `proxy.Direct`?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -129,22 +130,36 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+	switch proxyURL.Scheme {
+	case "socks5":
+		return s.dialWithSocks5Proxy(req, proxyURL)
+	case "https", "http":
```
from the docs for ProxyFromEnvironment:

> The environment values may be either a complete URL or a "host[:port]", in which case the "http" scheme is assumed.

should "" be included in this case as well?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -129,22 +130,36 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
 		Host:   targetHost,
 	}
 
+	proxyReq.WithContext(req.Context())
```
this needs to do `proxyReq = proxyReq.WithContext(req.Context())` to be effective

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -157,7 +172,46 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+		return nil, err
+	}
+
+	proxyDialConn, err := proxyDialer.Dial("tcp", targetHost)
```
I don't see us passing a context into here or into `proxy.SOCKS5`... what ensures this dial is cancelled if the parent context times out or is cancelled?

liggitt replied:
it looks like the value returned from `proxy.SOCKS5` implements `ContextDialer` ... should we expect/use that and plumb in the `req.Context()`?

TGPSKI replied:
Solved:

`proxyDialConn, err := proxyDialer.(proxy.ContextDialer).DialContext(req.Context(), "tcp", targetHost)`

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -314,6 +363,47 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+						proxyHandler = socks5Server(&socks5.StaticCredentials{
+							"proxyuser": "proxypasswd",
+						})
+						proxyCalledWithAuth = true
```
this seems misleading... this is setting proxyCalledWithAuth when the proxy server hasn't even been called yet. compare with the way http proxy sets this inside the server handler based on the incoming request

liggitt replied:
setting a CredentialStore implementation that logs an auth attempt would match the http test better and would let us ensure the socks server was called, and that it was called with auth, and the content of the auth credentials

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -314,6 +363,47 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+
+					proxyCalledWithHost = "//127.0.0.1:" + strconv.Itoa(l.Addr().(*net.TCPAddr).Port)
+					go proxyHandler.Serve(l)
+					time.Sleep(100 * time.Millisecond)
```
is there a more robust way to make sure the server is ready before we continue?

TGPSKI replied:
@liggitt: do you have recommendations for an alternative strategy?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -314,6 +363,47 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+					defer l.Close()
+
+					proxyCalledWithHost = "//127.0.0.1:" + strconv.Itoa(l.Addr().(*net.TCPAddr).Port)
+					go proxyHandler.Serve(l)
```
does this shut down when the listener is closed?

TGPSKI replied:
@liggitt: how can I test this?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -373,11 +463,16 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 				}
-				if proxyCalledWithAuthHeader != expectedProxyAuth {
-					t.Fatalf("%s: Expected to see a call to the proxy with credentials %q, got %q", k, testCase.proxyAuth, proxyCalledWithAuthHeader)
+				if l != nil {
+					l.Close()
```
isn't the `l.Close()` unnecessary with the deferred close above?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -373,11 +463,16 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 				}
-				if len(expectedProxyAuth) == 0 && proxyCalledWithAuth {
-					t.Fatalf("%s: Proxy authorization unexpected, got %q", k, proxyCalledWithAuthHeader)
+				if testCase.socks5ProxyServerFunc == nil {
+					if len(expectedProxyAuth) == 0 && proxyCalledWithAuth {
+						t.Fatalf("%s: Proxy authorization unexpected, got %q", k, proxyCalledWithAuthHeader)
+					}
+					if proxyCalledWithAuthHeader != expectedProxyAuth {
+						t.Fatalf("%s: Expected to see a call to the proxy with credentials %q, got %q", k, testCase.proxyAuth, proxyCalledWithAuthHeader)
+					}
```
this change is sort of hard to follow... I would have thought the expectedProxyAuth/proxyCalledWithAuth and `proxyCalledWithAuthHeader != expectedProxyAuth` checks could have remained as-is if the socks5 test proxy server recorded the fact that it was called with auth and the specific auth it was called with similarly

liggitt commented:
> 3\. Where can I find a resource on adding a module to the k8s repo? I have spent a lot of time and have failed to correctly add the SOCKS5 proxy server module for testing. Is there a clear doc showing a walkthrough that I am missing?

sorry for the confusion, the docs for that are linked from the go.mod file:

https://github.com/kubernetes/kubernetes/blob/da54185c9eb076a71962ddb5aded24811467dbb3/go.mod#L1-L5

liggitt commented:
@kfox1111, thanks for pinging on slack ... sorry for the long delay, it was a combination of lack of expertise in the area and competing priorities. I think the test dependency is not great, but won't block on it.

I did an actual review pass on the functional changes and exercising tests. Let me know when this is ready for another look.

erikbryant commented:
@liggitt Thanks for the detailed review. We are looking at each of those items and hope to start working on them next week. :)

TGPSKI commented:
@liggitt thanks so much for the review comments. I will take a look and bring another pass soon. I'm getting my company to add this to my scheduled workload for next week.

TGPSKI commented:
We did not plan this ^^

kfox1111 commented:
closed?

TGPSKI commented:
> closed?

I'm working on it

kfox1111 commented:
Ah, ok. Thanks. :)

TGPSKI committed:
add SOCKS5 support for kubectl exec

TGPSKI committed:
fix formatting

TGPSKI committed:
fix test formatting, pull in example certificate updates

TGPSKI committed:
run ./hack/pin-dependency.sh github.com/armon/go-socks5 e753329

TGPSKI committed:
run ./hack/update-vendor.sh

TGPSKI committed:
add missing imports to roundtripper_test

TGPSKI committed:
add err = nil return to dialWithSocks5Proxy and dialWithoutProxy

TGPSKI committed:
add tls handling to dialWithHttpProxy

TGPSKI committed:
get tests to run, add tls handling to dialWithSocks5Proxy

TGPSKI committed:
handle http over socks5 case

TGPSKI commented:
@liggitt: I'm running into the same exact build problems related to adding a module. I followed the guide (`./hack/pin-dependency.sh github.com/armon/go-socks5 e753329` + `./hack/update-vendor.sh`), but the automated build server still returns errors. There's some explanation missing from the docs, or I am doing something wrong. Can you help?

```
 - /home/prow/go/src/github.com/kubernetes/kubernetes/WORKSPACE:54:1
ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/BUILD:9:1: no such package 'vendor/github.com/armon/go-socks5': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - /home/prow/go/src/github.com/kubernetes/kubernetes/vendor/github.com/armon/go-socks5 and referenced by '//staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy:go_default_test'
ERROR: Analysis of target '//staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy:go_default_test' failed; build aborted: no such package 'vendor/github.com/armon/go-socks5': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - /home/prow/go/src/github.com/kubernetes/kubernetes/vendor/github.com/armon/go-socks5
```

TGPSKI commented:
If anyone involved in this conversation has a solution for the automated build process errors, I still need help. The scripts to add modules still do not produce a working build. See the comment above ^^.
@kfox1111 @lavalamp @mikedanese @zioalex @ribbybibby @rezlam @vetch101 @gordonbondon

TGPSKI committed:
remove l.close()

liggitt commented:
running those two scripts on this branch produces the following diff for me, pushed as the last commit in 

> @liggitt: I'm running into the same exact build problems related to adding a module. I followed the guide (`./hack/pin-dependency.sh github.com/armon/go-socks5 e753329` + `./hack/update-vendor.sh`), but the automated build server still returns errors. There's some explanation missing from the docs, or I am doing something wrong. Can you help?

Interesting... it looks like test-only dependencies of submodules don't get vendored. Our main module tests apparently had 100% overlap with the test dependencies of our submodules before now. Doing this will let you work around the issue while we think about a more automated way to handle this:

```diff
diff --git a/build/tools.go b/build/tools.go
index 33958d7bfba..5fc2bed2677 100644
--- a/build/tools.go
+++ b/build/tools.go
@@ -16,10 +16,11 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// This package imports things required by build scripts, to force `go mod` to see them as dependencies
+// This package imports things required by build scripts and test packages of submodules, to force `go mod` to see them as dependencies
 package tools
 
 import (
+	// build script dependencies
 	_ "github.com/go-bindata/go-bindata/go-bindata"
 	_ "github.com/onsi/ginkgo/ginkgo"
 	_ "k8s.io/code-generator/cmd/go-to-protobuf"
@@ -29,4 +30,7 @@ import (
 	_ "k8s.io/gengo/examples/import-boss/generators"
 	_ "k8s.io/gengo/examples/set-gen/generators"
 	_ "k8s.io/kube-openapi/cmd/openapi-gen"
+
+	// submodule test dependencies
+	_ "github.com/armon/go-socks5" // for staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go
 )
```

TGPSKI committed:
patch build/tools.go to support vendoring test-only submodule dependencies

TGPSKI committed:
run ./hack/update-vendor.sh after build/tools.go patch to suppport vendoring test-only submodule dependency

TGPSKI commented:
@kfox1111 @lavalamp @mikedanese @zioalex @ribbybibby @rezlam @vetch101 @gordonbondon

Hey everyone! Great news: the nagging build problems with this PR look to be solved, thanks to @liggit. Much appreciated!

I need help to push this PR over the line. I am working on review comments from `roundtripper_test.go`.

# Issue 1: `armon/go-socks5` missing proxy.OnRequest() handler

Review comment: https://github.com/kubernetes/kubernetes/pull/84205/files#r448620096

I am trying to make the SOCKS5 tests look more like HTTP tests. When an HTTP/S test is encountered, an OnRequest() handler [is created](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go#L294). In the OnRequest handler, `proxyCalledWithHost`, `proxyCalledWithAuth`, and `proxyCalledWithAuthHeader` are set on a request. These variables are tested against expected values at the end of the loop. 

In order to make the SOCKS5 test look like HTTP/S, I need to be able to add a request handler that will set these values. Looking at `armon/go-socks5`, I see no obvious way to add a request handler to the server. Is there a method or workaround I am missing in order to accomplish this task?

# Issue 2: Verify SOCKS5 server is ready before continuing test

Review comment: https://github.com/kubernetes/kubernetes/pull/84205/files#r448620683

I see no exposed methods in `armon/go-socks5` that help with verifying that the server is ready to respond to requests. Does anyone have experience in this area? Does anyone know of a similar problem with a solution - launching a server as part of a test suite?

# Issue 3: Verify closing listener will shut down 

Review comment: https://github.com/kubernetes/kubernetes/pull/84205/files#r448620861

I understand this concept but don't know how to test it. As part of the test suite, how could you accomplish testing closing of the listener and ensuring that the server routine is shut down as well?

TGPSKI committed:
attempts at managing listener lifecycle

TGPSKI committed:
roundtripper_test: add interceptor for viewing socks5 authorization requests

TGPSKI committed:
debugging full test suite

TGPSKI committed:
add socks5 authentication interceptor struct, ensure goroutines do not fire fatalf after test return, all tests passing

TGPSKI commented:
/test pull-kubernetes-e2e-gce

TGPSKI committed:
rework socks5 listener close logic, include check for expected failed test

TGPSKI commented:
/retest

TGPSKI commented:
@liggit, ready for a review. I think this pass should address your remaining concerns, let me know what else I need to look at.

Major changes:

* Revised SOCKS5 server lisener + serveConn implementations
* Removed usage of time to wait for socks5.Serve, modified to use ServeConn + construct
* Added Interceptor struct which implements Rewrite, allowing us to see the auth method and username used for socks5 authentication
* Changed test logic to support usage of Interceptor for SOCKS5 tests in place of expectedProxyAuth

TGPSKI committed:
run hack/update-vendor.sh after rebase

TGPSKI committed:
use Errorf in goroutine

TGPSKI committed:
fix test race in socks5 serveconn close

TGPSKI commented:
/retest

TGPSKI committed:
remove continue

TGPSKI commented:
/retest

TGPSKI commented:
/test pull-kubernetes-bazel-test

liggitt commented:
the test failures look related to the PR:

```
=== RUN   TestRoundTripAndNewConnection/redirect_=_false
2020/07/29 22:06:22 http: TLS handshake error from 127.0.0.1:58072: remote error: tls: bad certificate
2020/07/29 22:06:23 [ERR] socks: Failed to authenticate: User authentication failed
==================
WARNING: DATA RACE
Read at 0x00c000076c50 by goroutine 154:
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy.TestRoundTripAndNewConnection.func1.8()
      staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:413 +0x196

Previous write at 0x00c000076c50 by goroutine 48:
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy.TestRoundTripAndNewConnection.func1()
      staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:309 +0x252c
  testing.tRunner()
      GOROOT/src/testing/testing.go:1108 +0x202

Goroutine 154 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy.TestRoundTripAndNewConnection.func1()
      staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:404 +0x4664
  testing.tRunner()
      GOROOT/src/testing/testing.go:1108 +0x202

Goroutine 48 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1159 +0x796
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy.TestRoundTripAndNewConnection()
      staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:57 +0x98
  testing.tRunner()
      GOROOT/src/testing/testing.go:1108 +0x202
```

issues with 1.19 and CI are still consuming my attention, I hope to get back to this next week

bendrucker commented:
I think this should be an easy one! The read is on `L413`:

```go
if err := proxyHandler.ServeConn(conn); err != nil && !testCase.shouldError {
```

And the write is `L309`:

```go
for k, testCase := range testCases {
```

So as `testCases` is iterated, `testCase` is written to, but is also read during the previous test's termination logic.

You can declare a new variable:

```go
for k, testCase := range testCases {
	testCase := testCase
```

Or you can pass the `testCase` object into the server goroutine.

https://github.com/golang/go/wiki/CommonMistakes

TGPSKI committed:
assign testCase := testCase at start of test loop

bendrucker commented:
You can also detect these races locally when you test:

https://golang.org/doc/articles/race_detector.html

TGPSKI commented:
/test pull-kubernetes-e2e-kind

TGPSKI commented:
@liggitt All green!

neolit123 commented:
/remove-sig cluster-lifecycle

mulatinho commented:
finally! thanks @TGPSKI for your work on this, I hope this get merged ASAP, I really want this feature :D

dswiecki commented:
:+1: for this feature

TGPSKI commented:
@liggit These multiple rebases are becoming a big time sink, when can we slot this PR for review? Would like to do a single rebase followed by a review if possible.

liggitt commented:
ack, wrapping up KEP reviews for 1.20, then this is on my short-list of PRs to complete.

cc @mikedanese for a second set of eyes on the roundtripper.go changes

mikedanese commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -159,6 +174,81 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+		return nil, err
+	}
+
+	proxyDialConn, err := proxyDialer.(proxy.ContextDialer).DialContext(req.Context(), "tcp", targetHost)
```
Why do we know this is going to be a ContextDialer?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -314,6 +379,59 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 					defer proxy.Close()
 				}
 
```
move `interceptor := &Interceptor{}` here to keep it local to each test so we don't get cross-test pollution, and pass it into `socks5ProxyServerFunc`

TGPSKI replied:
moved

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -314,6 +379,59 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+							"proxyuser": "proxypasswd",
+						})
+					} else {
+						proxyHandler = socks5Server(nil)
```
```suggestion
						proxyHandler = testCase.socks5ProxyServerFunc(nil, interceptor)
```

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -314,6 +379,59 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+					var l net.Listener
+					var proxyHandler *socks5.Server
+					if testCase.proxyAuth != nil {
+						proxyHandler = testCase.socks5ProxyServerFunc(&socks5.StaticCredentials{
+							"proxyuser": "proxypasswd",
+						})
```
```suggestion
						proxyHandler = testCase.socks5ProxyServerFunc(&socks5.StaticCredentials{
							"proxyuser": "proxypasswd",
						}, interceptor)
```

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -314,6 +379,59 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+						l.Close()
+					}()
+
+					proxyCalledWithHost = "//127.0.0.1:" + strconv.Itoa(l.Addr().(*net.TCPAddr).Port)
```
the point of this var is to capture the host the proxy handler saw, to make sure 1) the proxy was contacted and 2) the proxy was asked to proxy to the right place.

the 
```
if testCase.proxyServerFunc != nil {
```
check below needs to be expanded to 
```
if testCase.proxyServerFunc != nil || testCase.socks5ProxyServerFunc != nil {
```

Once that is done, the check will start failing because this is being hard-coded to the proxy's host, not the backend's host. Letting interceptor set proxyCalledWithHost would match how this was intended to be used.

This diff would make the socks5 tests work the same way as the http proxy ones (and change each iteration of the loop to its own isolated test so you could do something like:
```
go test ./vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy -run TestRoundTripAndNewConnection//SOCKS5`
```

```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go
index 55aa5f30807..8cb0e2de476 100644
--- a/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go
@@ -36,13 +36,16 @@ import (
 	"github.com/elazarl/goproxy"
 
 	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
 type Interceptor struct {
 	Authorization       socks5.AuthContext
+	proxyCalledWithHost *string
 }
 
 func (i *Interceptor) Rewrite(ctx context.Context, req *socks5.Request) (context.Context, *socks5.AddrSpec) {
+	*i.proxyCalledWithHost = req.DestAddr.Address()
 	i.Authorization = socks5.AuthContext(*req.AuthContext)
 	return ctx, req.DestAddr
 }
@@ -86,8 +89,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 				return ts
 			}
 
-			interceptor := &Interceptor{}
-			socks5Server := func(creds *socks5.StaticCredentials) *socks5.Server {
+			socks5Server := func(creds *socks5.StaticCredentials, interceptor *Interceptor) *socks5.Server {
 				var conf *socks5.Config
 				if creds != nil {
 					cator := socks5.UserPassAuthenticator{Credentials: creds}
@@ -96,7 +98,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 						Rewriter:    interceptor,
 					}
 				} else {
-					conf = &socks5.Config{}
+					conf = &socks5.Config{Rewriter: interceptor}
 				}
 
 				ts, err := socks5.New(conf)
@@ -108,7 +110,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 
 			testCases := map[string]struct {
 				serverFunc             func(http.Handler) *httptest.Server
-				socks5ProxyServerFunc  func(*socks5.StaticCredentials) *socks5.Server
+				socks5ProxyServerFunc  func(*socks5.StaticCredentials, *Interceptor) *socks5.Server
 				proxyServerFunc        func(http.Handler) *httptest.Server
 				proxyAuth              *url.Userinfo
 				clientTLS              *tls.Config
@@ -306,8 +308,9 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 				},
 			}
 
-			for k, testCase := range testCases {
-				testCase := testCase
+			for _, k := range sets.StringKeySet(testCases).List() {
+				testCase := testCases[k]
+				t.Run(k, func(t *testing.T) {
 					server := testCase.serverFunc(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 						if testCase.shouldError {
 							if e, a := httpstream.HeaderUpgrade, req.Header.Get(httpstream.HeaderConnection); e != a {
@@ -379,15 +382,16 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 						defer proxy.Close()
 					}
 
+					interceptor := &Interceptor{proxyCalledWithHost: &proxyCalledWithHost}
 					if testCase.socks5ProxyServerFunc != nil {
 						var l net.Listener
 						var proxyHandler *socks5.Server
 						if testCase.proxyAuth != nil {
 							proxyHandler = testCase.socks5ProxyServerFunc(&socks5.StaticCredentials{
 								"proxyuser": "proxypasswd",
-						})
+							}, interceptor)
 						} else {
-						proxyHandler = socks5Server(nil)
+							proxyHandler = testCase.socks5ProxyServerFunc(nil, interceptor)
 						}
 
 						closed := false
@@ -400,8 +404,6 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 							l.Close()
 						}()
 
-					proxyCalledWithHost = "//127.0.0.1:" + strconv.Itoa(l.Addr().(*net.TCPAddr).Port)
-
 						go func() {
 							for {
 								conn, err := l.Accept()
@@ -422,13 +424,11 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 
 						spdyTransport.proxier = func(proxierReq *http.Request) (*url.URL, error) {
 							proxierCalled = true
-						proxyURL, err := url.Parse(proxyCalledWithHost)
-						if err != nil {
-							return nil, err
-						}
-						proxyURL.User = testCase.proxyAuth
-						proxyURL.Scheme = "socks5"
-						return proxyURL, nil
+							return &url.URL{
+								Scheme: "socks5",
+								Host:   net.JoinHostPort("127.0.0.1", strconv.Itoa(l.Addr().(*net.TCPAddr).Port)),
+								User:   testCase.proxyAuth,
+							}, nil
 						}
 					}
 
@@ -444,7 +444,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 						t.Fatalf("%s: shouldError=%t, got %t: %v", k, e, a, err)
 					}
 					if testCase.shouldError {
-					continue
+						return
 					}
 					defer conn.Close()
 
@@ -477,7 +477,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 						t.Fatalf("%s: expected '%s', got '%s'", k, e, a)
 					}
 
-				if testCase.proxyServerFunc != nil {
+					if testCase.proxyServerFunc != nil || testCase.socks5ProxyServerFunc != nil {
 						if !proxierCalled {
 							t.Fatalf("%s: Expected to use a proxy but proxier in SpdyRoundTripper wasn't called", k)
 						}
@@ -513,6 +513,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 							t.Fatalf("%s: Expected to see a call to the proxy with credentials %q, got %q", k, testCase.proxyAuth, proxyCalledWithAuthHeader)
 						}
 					}
+				})
 			}
 		})
 	}
```

TGPSKI replied:
Thanks for the detailed review! I will address these comments and come back with a code update.

TGPSKI replied:
I'm currently stuck trying to figure out how to access the interceptor methods outside of the `if testCase.socks5ProxyServerFunc != nil {` closure. Any hints you could provide?

TGPSKI replied:
Based on the review comments, I assume that no changes were needed to the final set of test checks (lines 495-515). I am unsure if your intention was for this section to be removed entirely or not. Because no removal suggestions were included above, I assume that this logic should remain.

liggitt replied:
notice in my example the `interceptor` variable is a concrete type:

`interceptor := &Interceptor{proxyCalledWithHost: &proxyCalledWithHost}`

you should be able to access the member fields of the concrete interceptor test implementation during the verification step

TGPSKI commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -126,6 +127,18 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+	case "socks5":
+		return s.dialWithSocks5Proxy(req, proxyURL)
+	case "https", "http", "":
+		return s.dialWithHttpProxy(req, proxyURL)
```
Is this comment intended to say that the switch statement is not needed, and we can combine dialWithSocks5Proxy and dialWithHttpProxy to a single method, dialWithHttpProxy?

TGPSKI replied:
I am unsure what the intent of the comment is.

TGPSKI commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -159,6 +174,81 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+
+	dialer := s.Dialer
+	if dialer == nil {
+		dialer = &net.Dialer{}
```
Adding the dialer options here for consistency

TGPSKI committed:
add dialer options from DefaultTransport

TGPSKI committed:
remove rebase messages from vendor/modules.txt

TGPSKI committed:
clean up rebase

TGPSKI committed:
clean up rebase pt 2

TGPSKI committed:
move interceptor creation to test loop

TGPSKI committed:
run hack/update-vendor.sh after rebase

xaocon commented:
Thanks for all this hard work @TGPSKI et all. Really happy this will be around.

TGPSKI committed:
add canonicalAddr, portMap, idnaASCII, and isASCII to support default port mapping for SOCKS5 addrresses

TGPSKI committed:
run update-vendor.sh to add idna package for canonicalAddr

TGPSKI commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -159,6 +174,81 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
+func (s *SpdyRoundTripper) dialWithSocks5Proxy(req *http.Request, proxyURL *url.URL) (net.Conn, error) {
+	// ensure we use a canonical host with proxyReq
+	targetHost := netutil.CanonicalAddr(req.URL)
+	proxyDialAddr := netutil.CanonicalAddr(proxyURL)
```
Added port mappings and support functions in these two commits:

https://github.com/kubernetes/kubernetes/pull/84205/commits/fff298f62e3ede04b5176148e30ac21bc5282001
https://github.com/kubernetes/kubernetes/pull/84205/commits/b73c4bd035959d8475680b4dbd6ae6b94f75bd72

TGPSKI committed:
cast testCase set items to local test case, use t.Run to separate test runs, and have proxier return scheme, host, and user per test case

TGPSKI committed:
cast testCase set items to local test case, use t.Run to separate test runs, and have proxier return scheme, host, and user per test case

TGPSKI commented:
Hi, I am looking for assistance implementing the changes requested by @liggitt (https://github.com/kubernetes/kubernetes/pull/84205#discussion_r509326938).

His goal was to make socks5 validation tests look the same as http / https proxy validation tests.

I've implemented his changes, but I am now stuck attempting to use Interceptor to capture the authorization context on a per-test basis.

Previously, an interceptor was cast at the start of the test loop and was part of the test loop closure. Now, intercepter is cast when `testCase.socks5ProxyServerFunc != nil`, putting the interceptor variable into a closure that is not accessible during the end of loop validity tests.

Does anyone have any idea how I can access the returned value from `interceptor.GetAuthContext()` outside of the closure scope where interceptor is defined?

TGPSKI commented:
Quick update: still seeking help as outlined in my post from late November: https://github.com/kubernetes/kubernetes/pull/84205#issuecomment-732479593

liggitt commented:
> Does anyone have any idea how I can access the returned value from `interceptor.GetAuthContext()` outside of the closure scope where interceptor is defined?

from https://github.com/kubernetes/kubernetes/pull/84205#discussion_r529136326

> notice in my example the interceptor variable is a concrete type:
> 
> interceptor := &Interceptor{proxyCalledWithHost: &proxyCalledWithHost}
> 
> and since it is constructed outside the if testCase.socks5ProxyServerFunc != nil block, you should be able to access the member fields of the concrete interceptor test implementation during the verification step

liggitt commented:
feel free to reassign once ready for review again

TGPSKI commented:
Thanks @liggitt!

kfox1111 commented:
Any movement on this? Not having socks support is painful for some of us users.

kfox1111 commented:
/remove-lifecycle rotten

xens commented:
@TGPSKI if you could give us as status of what still needs to be done we can probably help you to finish the work. Based on your latest comment where your were seeking for help (https://github.com/kubernetes/kubernetes/pull/84205#issuecomment-732479593), did you do any progress on the interceptor thing or should we rework that part from scratch ?

terinjokes commented:
When I reached out in email a few weeks ago, it seemed like the two main things were to rebase everything on top on master again then working on the interceptor. I've started on this, but I've been interrupted by other work tasks.

xens commented:
Awesome, let me know how it goes, happy to help if there's anything to do.

nlamirault commented:
/remove-lifecycle rotten

xens commented:
I think we can close this one in favor of https://github.com/kubernetes/kubernetes/pull/105632
